src/main/java/org/terasology/metalrenegades/ai/actions/NearbyTargetInsideMarketAction.java

```diff
@@ -0,0 +1,103 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.metalrenegades.ai.actions;
+
+import com.google.common.collect.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.dynamicCities.buildings.components.DynParcelRefComponent;
+import org.terasology.dynamicCities.parcels.DynParcel;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.logic.behavior.BehaviorAction;
+import org.terasology.logic.behavior.core.Actor;
+import org.terasology.logic.behavior.core.BaseAction;
+import org.terasology.logic.behavior.core.BehaviorState;
+import org.terasology.math.geom.Rect2i;
+import org.terasology.minion.move.MinionMoveComponent;
+
+import java.util.List;
+import java.util.Random;
+import org.terasology.metalrenegades.ai.component.HomeComponent;
+import org.terasology.navgraph.WalkableBlock;
+
+/**
+ * Sets a marketkeeper character's {@link MinionMoveComponent} target to a random nearby block inside the area of
+ * their market building.
+ */
+@BehaviorAction(name = "set_target_inside_market")
+public class NearbyTargetInsideMarketAction extends BaseAction {
+
+    private static final Logger logger = LoggerFactory.getLogger(NearbyTargetInsideMarketAction.class);
+
+    private static final int RANDOM_BLOCK_ITERATIONS = 10;
+    private Random random = new Random();
+
+    private Rect2i marketRegion;
+
+    private int moveProbability = 100;
+
+    @Override
+    public void construct(Actor actor) {
+        HomeComponent homeComp = actor.getComponent(HomeComponent.class);
+        EntityRef buildingEntity = homeComp.building;
+
+        DynParcelRefComponent dynParcelRefComponent = buildingEntity.getComponent(DynParcelRefComponent.class);
+        DynParcel dynParcel = dynParcelRefComponent.dynParcel;
+
+        marketRegion = dynParcel.getShape();
+    }
+
+    @Override
+    public BehaviorState modify(Actor actor, BehaviorState result) {
+        if (random.nextInt(100) > (99 - moveProbability)) {
+            MinionMoveComponent moveComponent = actor.getComponent(MinionMoveComponent.class);
+            if (moveComponent.currentBlock != null) {
+                WalkableBlock target = findRandomNearbyBlockInMarket(moveComponent.currentBlock);
+                moveComponent.target = target.getBlockPosition().toVector3f();
+                actor.save(moveComponent);
+            } else {
+                return BehaviorState.FAILURE;
+            }
+        }
+        return BehaviorState.SUCCESS;
+    }
+
+    /**
+     * Finds a block close to the character, with the condition that this block must be inside the market parcel.
+     *
+     * @param startBlock The block that this characer is currently standing on.
+     * @return A random nearby block nearby inside the market area.
+     */
+    private WalkableBlock findRandomNearbyBlockInMarket(WalkableBlock startBlock) {
```

Comment:
May I ask why you approached it this way? Sure, it's definitely more robust considering the navgraphSystem hiccups we might come across but it sort of seems more complicated than it could be. Did you try just finding a random block within the lot of the market? and doing a getBlock() on that?
`WalkableBlock nextBlock = pathfinderSystem.getBlock(randomPos)`

Reply:
This node is made to be very similar to the way that the original stray behavior worked, the node here is basically this: https://github.com/Terasology/Behaviors/blob/0928e657b30b8ef2b1fadf6e718c697c8f9f22cc/src/main/java/org/terasology/minion/move/SetTargetToNearbyBlockNode.java with market size limitations considered.

Choosing random blocks in the marketplace would probably be much simpler, is this the way that I should go?